src/pages/SixCoinPuzzle.js

```diff
@@ -0,0 +1,131 @@
+import "../styles.css";
+import React, { useState } from "react";
+import toast, { Toaster } from "react-hot-toast";
+import ArrowKeysReact from "arrow-keys-react";
+import Button from "../components/Button";
+import Coin from "../components/Coin";
+
+const SixCoinPuzzle = () => {
+  const numberOfCoinSlots = 7;
+  const startingArray = [1, 1, 1, 0, 2, 2, 2];
+  const winningArray = [2, 2, 2, 0, 1, 1, 1];
+  const [coinTypeArray, setCoinTypeArray] = useState(startingArray);
+  const [selectedCoin, setSelectedCoin] = useState(0);
+  const [moveCounter, setMoveCounter] = useState(0);
+  const [coinSelectText, setCoinSelectText] = useState(
+    "Click to select a coin"
+  );
+
+  const swapCoins = (array, index1, index2) => {
+    [array[index1], array[index2]] = [array[index2], array[index1]];
+    setCoinTypeArray(array);
+    selectCoin(index2);
+    setMoveCounter(moveCounter + 1);
+    checkCompletion(array);
+  };
+  const selectCoin = (coinNumber) => {
+    // cannot select empty coins
+    if (coinTypeArray[coinNumber] !== 0) {
+      setSelectedCoin(coinNumber);
+      console.log("selected coin: ", { coinNumber });
+      setCoinSelectText("Use arrow keys to move");
+    }
+  };
+  const handleMoveLeft = () => {
+    const leftCoin = selectedCoin - 1;
+    const jumpLeftCoin = selectedCoin - 2;
+    let temp = coinTypeArray;
+    // first check that coin is eligible to move left, then either move one space left
+    // or jump two spaces left
+    if (coinTypeArray[selectedCoin] === 1) {
+      toast.error("cannot move silver coin left", {
+        position: "bottom-center",
+      });
+    } else if (coinTypeArray[leftCoin] === 0) {
+      swapCoins(temp, selectedCoin, leftCoin);
+    } else if (coinTypeArray[jumpLeftCoin] === 0) {
+      swapCoins(temp, selectedCoin, jumpLeftCoin);
+    } else {
+      toast.error("no available space", { position: "bottom-center" });
+    }
+  };
+  const handleMoveRight = () => {
+    const rightCoin = selectedCoin + 1;
+    const jumpRightCoin = selectedCoin + 2;
+    let temp = coinTypeArray;
+    // first check that coin is eligible to move right, then either move one space right
+    // or jump two spaces right
+    if (coinTypeArray[selectedCoin] === 2) {
+      toast.error("cannot move gold coin right", {
+        position: "bottom-center",
+      });
+    } else if (coinTypeArray[rightCoin] === 0) {
+      swapCoins(temp, selectedCoin, rightCoin);
+    } else if (coinTypeArray[jumpRightCoin] === 0) {
+      swapCoins(temp, selectedCoin, jumpRightCoin);
+    } else {
+      toast.error("no available space", { position: "bottom-center" });
+    }
+  };
+  const checkCompletion = (array) => {
+    if (array.every((val, index) => val === winningArray[index])) {
+      toast.success("Congrats! You've completed the puzzle");
+    }
+  };
+  const resetPuzzle = () => {
+    setCoinTypeArray(startingArray);
+    setMoveCounter(0);
+    setCoinSelectText("Click to select a coin");
```

Comment:
Instead of setting coin text as state, another option is to have a status state variable, and essentially treat the whole component as a finite state machine. Then things like coin text can be derived from the top level status.

```
const [status, setStatus] = useState('Idle')

// useCallback declares a dependency on status
// This ensures that if status changes, this function will always have the latest
// status value when it evaluates.
// If we did not do this, then we could end up with a stale status value inside
// of the function due to the way JavaScript closures work.
const getCointText = useCallback(() => {
    switch (status) {
        case 'Idle':
            return 'Click to select a coin'
        case 'Selected':
            return 'Use arrow keys to move'
    }
}, [status])

return (
    <div className="coin-select-text">
        <p>{getCoinText}</p>
    </div>
)
```

This is really useful because instead of having to call setCointText and manually pass in the coin text in multiple places, you can just change the status for the component, and coin text and possibly other bits of UI can update accordingly.